xbe-loader/main.c

```diff
@@ -0,0 +1,429 @@
+  unsigned int numberOfBytesRead;
+  XReadFile(f->handle, buffer, chunk_count * chunk_size, &numberOfBytesRead);
+  return numberOfBytesRead / chunk_size;
+}
+
+int fseek(FILE* f, int offset, int whence) {
+  int moveMethod;
+  switch(whence) {
+  case SEEK_SET:
+    moveMethod = FILE_BEGIN;
+    break;
+  case SEEK_END:
+    moveMethod = FILE_END;
+    offset = -offset;
+    break;
+  default:
+    debugPrint("Bad whence: %d\n", whence);
+    while(1);
+  }
+
+  int newFilePointer;
+  XSetFilePointer(f->handle, offset, &newFilePointer, moveMethod);
+
+  return 0;
+}
+
+int fopen(char* path, char* mode) {
+  int handle;
+  NTSTATUS status = XCreateFile(&handle, path, GENERIC_READ, 0, OPEN_EXISTING, 0);
+
+  // Error out if no file was loaded
+  if (status != STATUS_SUCCESS) {
+    return NULL;
+  }
+
+  // Create a FILE object
+  FILE* f = malloc(sizeof(FILE));
+  f->handle = handle;
+
+  //FIXME: Might be implied? Better safe than sorrty for now
+  fseek(f, 0, SEEK_SET);
+
+  return f;
+}
+
+int fclose(FILE* f) {
+  XCloseHandle(f->handle);
+}
+
+#endif
+
+
+
+// Section will be copied to new versions of loader
+#define __PERSIST __attribute__ ((section ("persist")))
+
+// Section will be removed on first copy
+#define __INIT __attribute__ ((section ("init")))
+
+#define __NO_RETURN
+
+#define DEFAULT_BASE 0x10000
+
+// PE relocation types
+#define IMAGE_REL_BASED_HIGHLOW              3
+
+const uint32_t XOR_EP_DEBUG                            = 0x94859D4B; // Entry Point (Debug)
+const uint32_t XOR_EP_RETAIL                           = 0xA8FC57AB; // Entry Point (Retail)
+const uint32_t XOR_KT_DEBUG                            = 0xEFB1F152; // Kernel Thunk (Debug)
+const uint32_t XOR_KT_RETAIL                           = 0x5B6D40B6; // Kernel Thunk (Retail)
+
+typedef struct {
+  uint32_t flags;
+  uint32_t virtual_address;
+  uint32_t virtual_size;
+  uint32_t raw_address;
+  uint32_t raw_size;
+  uint32_t name;
+  uint32_t ref_counts[3]; //FIXME: These are pointers I think
+  uint8_t hash[20];
+} Section;
+
+void* loader_path __PERSIST = NULL;
+void* loader_unload __PERSIST = NULL;
+uint32_t loader_base_address __PERSIST = DEFAULT_BASE;
+
+
+static uint32_t LookupKernelExport(unsigned int ordinal) {
+  uint32_t image_base = 0x80010000;
+  uint32_t tmp = *(uint32_t*)(image_base + 0x3C);
+  tmp = *(uint32_t*)(image_base + tmp + 0x78);
+  //ExportCount = read_u32(image_base + TempPtr + 0x14);
+  uint32_t ExportBase = image_base + *(uint32_t*)(image_base + tmp + 0x1C);
+  //#FIXME: Read all exports at once and parse them locally
+
+  //#for i in range(0, ExportCount):
+  //#  ordinal = i + 1
+  //#  print("@" + str(ordinal) + ": 0x" + format(image_base + read_u32(ExportBase + i * 4), '08X'))
+
+  //index = (ordinal - 1) # Ordinal
+  //#assert(index < ExportCount) #FIXME: Off by one?
+
+  unsigned int index = ordinal - 1;
+
+  return image_base + *(uint32_t*)(ExportBase + index * 4);
```

Comment:
Perhaps calculate `ExportBase` once in caller, so that `LookupKernelExport` can accept is as an argument?

Reply:
true